### PR TITLE
Feature/my enumerables

### PR DESCRIPTION
**Pull request by ebeagusamuel**
#### What does this PR do?
- This PR is an implementation of my own version of some of Ruby's enumerable method.
1. **#my_each method**
  Implementation of [`#each`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-each_with_index)
2. **#my_each_with_index method**
  Implementation of [`#each_with_index`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-each_with_index)
3. **#my_select (#my_find_all, #my_filter) method**
  Implementation of [`#select, find_all, filter`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-select)
4. **#my_all? method**
  Implementation of [`#all?`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-all-3F)
5. **#my_any? method**
  Implementation of [`#any?`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-any-3F)
6. **#my_none? method**
  Implementation of [`#none?`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-none-3F)
7. **#my_count method**
  Implementation of [`#count`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-count)
8. **#my_map method**
   Customize Implementation of [`#map`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-map). this method will now take `Proc` as an argument unlike the original one. if both `Proc` and `block` is given `Proc` will have precedence
9. **#my_inject (#my_reduce) method**
  Implementation of [`#inject, (#reduce)`](https://ruby-doc.org/core-2.7.1/Enumerable.html#method-i-inject)
10. **#multiply_els method**
   New method only created for testing `my_inject` method

### CODE REVIEW

File: my_enumerables.rb
Comment (stratospherique) [inline my_enumerables.rb]: * [ ] should have working `my_each`identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [ ] my_each returns an Enumerator if no block is given
  ```ruby
    array = [1,2,5,8,7,98,4]
    p array.my_each.class # not Enumerator
  ```
* [ ] my_each when a block is given after calling the given block once for each element in self, returns the array itself
  ```ruby
    array = [1,2,5,8,7,98,4]
    block = proc { |num| num < 8 }
    p array.my_each(&block) == array.each(&block) # false
  ```

File: my_enumerables.rb
Comment (stratospherique) [inline my_enumerables.rb]: * [ ] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [ ] my_each_with_index returns an enumerator if no block is given
  ```ruby
    array = [1,2,5,8,7,98,4]
    p array.my_each_with_index.class # not Enumerator
  ```

* [ ] my_each_with_index when a block is given after calling the given block once for each element in self, returns the array itself
   ```ruby
    array = [1,2,5,8,7,98,4]
    block = proc { |num| num < 8 }
    p array.my_each_with_index(&block) == array.each_with_index(&block) # false
  ```

File: my_enumerables.rb
Comment (stratospherique) [inline my_enumerables.rb]: * [ ] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_select returns an enumerator if no block is given
  ```
    array = [1,2,5,8,7,98,4]
    p array.my_select.class # not Enumerator
  ```

File: my_enumerables.rb
Comment (stratospherique) [inline my_enumerables.rb]: * [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [ ] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map #error
    # LocalJumpError:
        no block given (yield)
      # ./my_enumerables.rb:150:in `block in my_map'
      # ./my_enumerables.rb:13:in `block in my_each'
      # ./my_enumerables.rb:12:in `upto'
      # ./my_enumerables.rb:12:in `my_each'
      # ./my_enumerables.rb:149:in `my_map'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great start @ebeagusamuel :clap: :clap: However, some methods need a bit of refactoring in order to get approval for this project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods :gear: 
#### my_all?, my_any?, my_none?
The issues with **my_none?**, **my_any?** and **my_all?** are similar. the methods should be able to: 

* [ ] Return a suitable boolean result if no arguments were passed.
  ```ruby
	array = [false, true, false, nil, []]
	puts array.my_any? # should print true
	puts array.my_all? # should print false
	puts array.my_none? # should print false
  ```
* [ ] Accept primitives, classes (i.e Integer, String, etc...) and Regex Expressions as arguments.
  ```ruby
	words = %w[dog door rod blade]
	puts words.my_all?(/c/) == words.all?(/c/) # false
	puts words.my_any?(/c/) == words.all?(/c/) # false
	puts words.my_none?(/c/) == words.all?(/c/) # true
  ```

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by izaiasneto4** [comment]
### Hi @ebeagusamuel, nicely done! :clap: There's just one more thing that we need to fix before we can get your project approved :arrow_down: 

- [ ] In the following example given by @stratospherique :arrow_down: 
```ruby
array = [false, true, false, nil, []]

puts array.my_none? # should print false
```
Your method should output `false` with the example, but it still returns true :wink: 

Please, when finished fixing this little error re-submit the project for a new review :pray: 

_TSE: Izaias Neto_

**Review by Forison — APPROVED** [review]
## status:: approved✔️
# 🙋🏽‍♂️ @ebeagusamuel,
Good job on addressing all the issues raised by my colleagues earlier🙇🏻
## Go ahead and merge the branch and proceed to the next task✈️
#### Happy coding
_Addo

**Pull request merged**
